server/services/gameTick.ts

```diff
@@ -323,102 +327,261 @@ export default class GameTickService extends EventEmitter {
+            for (let i = 0; i < positions.length - 1; i++) {
                 let friendlyCarrier = positions[i];
 
-                if (friendlyCarrier.carrier.ships <= 0) {
+                // Checks if the carrier is still alive and not a gift.
+                if (friendlyCarrier.carrier.ships <= 0 || friendlyCarrier.carrier.isGift) {
+                    continue;
+                }
+
+                // Check if it has a specialist that avoids c2cc.
+                if (this._checkAvoidCarrierToCarrierCombat(friendlyCarrier.carrier)) {
                     continue;
                 }
 
-                // First up, get all carriers that are heading from the destination and to the source
-                // and are in front of the carrier.
-                let collisionCarriers: CarrierPosition[] = positions
-                    .filter((c: CarrierPosition) => {
-                        return (c.carrier.ships! > 0 && !c.carrier.isGift) // Is still alive and not a gift
+                for (let j = i + 1; j < positions.length; j++) {
+                    let enemyCarrier = positions[j];
+
+                    // Checks if a carrier is alive and not a gift, because only then can it engage in c2cc.
+                    if (enemyCarrier.carrier.ships <= 0 || enemyCarrier.isGift) {
+                        continue;
+                    }
+
+                    // Check if carriers have the same owner, if so, no combat occurs.
+                    if (friendlyCarrier.carrier.ownedByPlayerId === enemyCarrier.carrier.ownedByPlayerId) {
+                        continue;
+                    }
+
+                    // Check if the players are allied, in which case no combat occurs.
+                    if (isAlliancesEnabled) {
+                        const playerIds: DBObjectId[] = [friendlyCarrier.carrier.ownedByPlayerId, enemyCarrier.carrier.ownedByPlayerId];
+
+                        const isAllPlayersAllied = this.diplomacyService.isDiplomaticStatusBetweenPlayersAllied(game, playerIds);
+
+                        if (isAllPlayersAllied) {
+                            continue;
+                        }
+                    }
+
+                    let friendlySpeed = friendlyCarrier.actualSpeed;
+                    let enemySpeed = enemyCarrier.actualSpeed;
+                    let friendlyDistance = friendlyCarrier.distanceToDestinationCurrent - friendlyCarrier.distanceToDestinationNext;
+                    let enemyDistance =  enemyCarrier.distanceToDestinationCurrent - enemyCarrier.distanceToDestinationNext;
+
+                    // Checks if carriers will actually cross paths.
+                    let willCrossPaths = (
+                        // Head-to-head combat
+                        (
+                            enemyCarrier.destination.toString() === friendlyCarrier.source.toString()
+                            && enemyCarrier.distanceToSourceCurrent < friendlyCarrier.distanceToDestinationCurrent + 10**-6
```

Comment:
`10**-6` - Is this to prevent rounding errors?

Reply:
Exactly. Normally carriers would follow the exact same paths and therefore have the exact same rounding errors. But since some carriers travel at different speeds it means their rounding errors can also slightly differ. Meaning you have to allow a certain amount of wiggling room.